pkg/telemetry/span.go

```diff
@@ -0,0 +1,555 @@
+		tid, sid = gen.NewIDs(ctx)
+	} else {
+		sid = gen.NewSpanID(ctx, tid)
+	}
+	// Take grantparent span to override psc's spanID
+	if so.OverrideParentSpanID() {
+		sid = psc.SpanID()
+		pid := so.ParentSpanID()
+		psc = psc.WithSpanID(*pid)
+	}
+	if so.OverrideSpanID() {
+		sid = *so.SpanID()
+
+		if so.NewRoot() {
+			psc = trace.SpanContext{}
+		} else if so.OverrideParentSpanID() {
+			pid := so.ParentSpanID()
+			psc = psc.WithSpanID(*pid)
+		}
+	}
+
+	sconf := trace.SpanContextConfig{
+		TraceID:    tid,
+		SpanID:     sid,
+		TraceState: psc.TraceState(),
+		TraceFlags: psc.TraceFlags(),
+	}
+
+	s := &Span{
+		parent:   psc,
+		name:     so.SpanName(),
+		scope:    instrumentation.Scope{Name: so.SpanScope()},
+		resource: so.Resource(),
+		start:    so.Timestamp(),
+		attrs:    so.Attributes(),
+		events:   []tracesdk.Event{},
+		links:    so.Links(),
+		status:   tracesdk.Status{Code: codes.Unset},
+		conf:     sconf,
+		kind:     so.SpanKind(),
+		dedup:    so.Dedup(),
+	}
+
+	return trace.ContextWithSpan(ctx, s), s
+}
+
+// span is an attempt to mimic the otel span data structure following the protobuf spec at
+// https://github.com/open-telemetry/opentelemetry-proto/blob/v1.1.0/opentelemetry/proto/trace/v1/trace.proto
+//
+// Due to the limitations of the otel lib's API interface, we can't reconstruct spans over boundaries,
+// and in order to make sure the execution data looks like how it looks from the SDK side,
+// we'll need to work around the otel library and have slightly different way of working with the data
+//
+// This file is an attempt to make it as close as possible to official libs so we can minimize deviations.
+//
+// NOTE: to make sure it doesn't conflict the the ReadOnlySpan interface functions,
+// certain fields are named in a little weird way.
+type Span struct {
+	tracesdk.ReadWriteSpan // embeds both span interfaces
+	sync.Mutex
+
+	start time.Time
+
+	name   string
+	attrs  []attribute.KeyValue
+	status tracesdk.Status
+	events []tracesdk.Event
+	links  []tracesdk.Link
+	kind   trace.SpanKind
+
+	scope    instrumentation.Scope
+	resource *resource.Resource
+	parent   trace.SpanContext
+	conf     trace.SpanContextConfig
+
+	// dedup marks the span as a potential duplicate, which in turn
+	// can be used as an indicator to allow the span to be sent of not.
+	dedup bool
+	// Mark the span as cancelled, so it doesn't get sent out when it ends
+	cancel            bool
+	childSpanCount    int
+	droppedAttributes int
+}
+
+// Send is just an alias for End
+func (s *Span) Send() {
+	s.End()
+}
+
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+//
+//	FOOTGUN ALERT
+//
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// Cancel will mark the span as cancelled and it will not be sent out when it ends.
+// This will reset the context so if there are spans that will be created after this,
+// it doesn't create a dangling pointer.
+func (s *Span) Cancel(ctx context.Context) context.Context {
```

Comment:
Need a better way to highlight this is a footgun that we need to be careful with.